File: src/lib/conversations/mappers.ts
/**
 * Conversation mappers — convert Prisma rows to DTOs.
 * Centralized so all conversation API routes return consistent shapes.
 */

import type {
  ConversationSummary,
  ConversationDetail,
  ConversationMessage,
  ConversationChange,
  ConversationParticipant,
} from './types'
import { detectMessageType } from './resolver'
import type { Conversation, Email, CategoryMembership, Category, Thread } from '@prisma/client'

/** Parse participant summary JSON safely. */
function parseParticipants(json: string): ConversationParticipant[] {
  try {
    return JSON.parse(json || '[]') as ConversationParticipant[]
  } catch {
    return []
  }
}

/** Map a Conversation row (with includes) to a ConversationSummary. */
export function mapConversationSummary(
  c: Conversation & {
    threads?: Thread[]
    _count?: { threads: number }
  },
): ConversationSummary {
  const participants = parseParticipants(c.participantSummary)
  const threadCount = c._count?.threads ?? c.threads?.length ?? 0
  return {
    id: c.id,
    accountId: c.accountId,
    canonicalSubject: c.canonicalSubject,
    status: c.status as ConversationSummary['status'],
    followUpState: c.followUpState as ConversationSummary['followUpState'],
    importance: c.importance as ConversationSummary['importance'],
    messageCount: c.messageCount,
    threadCount,
    participants,
    latestMessageAt: c.latestMessageAt?.toISOString() ?? null,
    firstMessageAt: c.firstMessageAt?.toISOString() ?? null,
    latestSubject: null,
    latestFromEmail: null,
    unreadCount: 0,
    hasAttachments: false,
    hasDeadline: false,
    categoryId: null,
    categoryName: null,
    categoryColor: null,
  }
}

/** Map a full conversation with messages + changes to ConversationDetail. */
export function mapConversationDetail(
  c: Conversation & {
    threads: (Thread & {
      emails: (Email & {
        memberships: (CategoryMembership & { category: Category })[]
      })[]
    })[]
  },
): ConversationDetail {
  const base = mapConversationSummary(c)

  // Flatten all emails across threads, sorted by receivedAt
  const allEmails = c.threads
    .flatMap((t) => t.emails)
    .sort((a, b) => (a.receivedAt?.getTime() ?? 0) - (b.receivedAt?.getTime() ?? 0))

  const messages: ConversationMessage[] = allEmails.map((e, i) => {
    const cat = e.memberships[0]?.category
    return {
      id: e.id,
      emailId: e.id,
      threadId: e.threadId ?? null,
      providerThreadId: e.providerThreadId ?? null,
      fromName: e.fromName,
      fromEmail: e.fromEmail,
      subject: e.subject,
      snippet: e.snippet,
      receivedAt: e.receivedAt?.toISOString() ?? null,
      isRead: e.isRead,
      isImportant: e.isImportant,
      hasAttachment: e.hasAttachment,
      messageType: detectMessageType(e.subject, i, allEmails.length) as ConversationMessage['messageType'],
      isLatest: i === allEmails.length - 1,
      categoryId: cat?.id ?? null,
      categoryName: cat?.name ?? null,
      categoryColor: cat?.color ?? null,
    }
  })

  const changes = detectChanges(allEmails)

  return {
    ...base,
    messages,
    changes,
    aiSummary: c.aiSummary,
  }
}

/**
 * Detect changes across conversation messages (§12 "What changed?").
 * Extracts deadlines, eligibility, location, and other key fields from each
 * message and computes diffs. Grounded in actual message content — never fabricates.
 */
function detectChanges(
  emails: Email[],
): ConversationChange[] {
  const changes: ConversationChange[] = []
  if (emails.length < 2) return changes

  // Extract key fields from each email's body
  const fieldExtractors: { field: string; pattern: RegExp }[] = [
    { field: 'Deadline', pattern: /(?:deadline|due(?:\s+by)?|submit\s+by|last\s+date)\s*:?\s*([^\n.]{3,60})/i },
    { field: 'Eligibility', pattern: /(?:eligib(?:le|ility)|cgpa|grade)\s*:?\s*([^\n.]{3,60})/i },
    { field: 'Location', pattern: /(?:location|venue|auditorium|room)\s*:?\s*([^\n.]{3,60})/i },
    { field: 'Time', pattern: /(?:time|at)\s*:?\s*(\d{1,2}[:.]?\d{0,2}\s*(?:am|pm)?)/i },
    { field: 'Date', pattern: /(?:date|on)\s*:?\s*([^\n.]{3,40})/i },
  ]

  for (const { field, pattern } of fieldExtractors) {
    const values: { value: string; emailId: string; subject: string | null; receivedAt: Date | null }[] = []
    for (const e of emails) {
      const text = `${e.subject ?? ''}\n${e.bodyText ?? ''}`
      const match = text.match(pattern)
      if (match?.[1]) {
        values.push({
          value: match[1].trim().slice(0, 80),
          emailId: e.id,
          subject: e.subject,
          receivedAt: e.receivedAt,
        })
      }
    }

    // If we have 2+ distinct values, record the change
    if (values.length >= 2) {
      const first = values[0]
      const last = values[values.length - 1]
      if (first.value.toLowerCase() !== last.value.toLowerCase()) {
        changes.push({
          field,
          earlier: first.value,
          latest: last.value,
          changedAt: last.receivedAt?.toISOString() ?? null,
          sourceMessageId: last.emailId,
          sourceSubject: last.subject,
        })
      }
    }
  }

  return changes
}

File: backend/app/domains/conversations/mappers.py
"""Prisma row → Pydantic DTO mappers for the conversations domain.

Mirrors `src/lib/conversations/mappers.ts`. ``detectMessageType`` and
``detectChanges`` port the TS heuristics so the FastAPI side returns the
same timeline + diff shape the frontend already consumes.
"""
from __future__ import annotations

import json
import re
from typing import Any

from app.domains.conversations.schemas import (
    ConversationChange, ConversationDetail, ConversationMessage,
    ConversationParticipant, ConversationSummary,
)

_FIELD_PATTERNS: list[tuple[str, re.Pattern[str]]] = [
    ("Deadline", re.compile(r"(?:deadline|due(?:\s+by)?|submit\s+by|last\s+date)\s*:?\s*([^\n.]{3,60})", re.I)),
    ("Eligibility", re.compile(r"(?:eligib(?:le|ility)|cgpa|grade)\s*:?\s*([^\n.]{3,60})", re.I)),
    ("Location", re.compile(r"(?:location|venue|auditorium|room)\s*:?\s*([^\n.]{3,60})", re.I)),
    ("Time", re.compile(r"(?:time|at)\s*:?\s*(\d{1,2}[:.]?\d{0,2}\s*(?:am|pm)?)", re.I)),
    ("Date", re.compile(r"(?:date|on)\s*:?\s*([^\n.]{3,40})", re.I)),
]


def _parse_participants(raw: str | None) -> list[ConversationParticipant]:
    """Parse the JSON-encoded participantSummary column (tolerant)."""
    if not raw:
        return []
    try:
        arr = json.loads(raw)
    except (TypeError, ValueError):
        return []
    if not isinstance(arr, list):
        return []
    out: list[ConversationParticipant] = []
    for item in arr:
        if isinstance(item, dict) and "email" in item:
            out.append(ConversationParticipant(
                email=item["email"], name=item.get("name"),
                role=item.get("role", "both"),
            ))
    return out


def detect_message_type(subject: str | None, index: int, total: int) -> str:
    """Classify a message within a thread (mirrors TS detectMessageType)."""
    if index == 0:
        return "original"
    s = (subject or "").lower()
    if index == total - 1:
        if "final" in s or "last" in s:
            return "final"
        if "reminder" in s:
            return "reminder"
        if "update" in s or "updated" in s:
            return "update"
    if "reminder" in s:
        return "reminder"
    if "follow" in s or "following up" in s:
        return "follow_up"
    if "clarif" in s or "question" in s:
        return "clarification"
    if "update" in s or "updated" in s or "revised" in s:
        return "update"
    return "reply"


def _detect_changes(emails: list[Any]) -> list[ConversationChange]:
    """Extract field diffs across messages (mirrors TS detectChanges)."""
    if len(emails) < 2:
        return []
    changes: list[ConversationChange] = []
    for field, pattern in _FIELD_PATTERNS:
        values: list[tuple[str, str, str | None, str | None]] = []
        for e in emails:
            text = f"{e.subject or ''}\n{e.bodyText or ''}"
            m = pattern.search(text)
            if m and m.group(1):
                values.append((
                    m.group(1).strip()[:80], e.id, e.subject,
                    e.receivedAt.isoformat() if e.receivedAt else None,
                ))
        if len(values) >= 2:
            first, last = values[0], values[-1]
            if first[0].lower() != last[0].lower():
                changes.append(ConversationChange(
                    field=field, earlier=first[0], latest=last[0],
                    changedAt=last[3], sourceMessageId=last[1],
                    sourceSubject=last[2],
                ))
    return changes


def map_conversation_summary(row: Any) -> ConversationSummary:
    """Project a Conversation row to the summary DTO."""
    thread_count = getattr(getattr(row, "_count", None), "threads", None) \
        or len(getattr(row, "threads", None) or [])
    return ConversationSummary(
        id=row.id, accountId=row.accountId,
        canonicalSubject=row.canonicalSubject, status=row.status,
        followUpState=row.followUpState, importance=row.importance,
        messageCount=row.messageCount, threadCount=thread_count,
        participants=_parse_participants(row.participantSummary),
        latestMessageAt=row.latestMessageAt.isoformat() if row.latestMessageAt else None,
        firstMessageAt=row.firstMessageAt.isoformat() if row.firstMessageAt else None,
        latestSubject=None, latestFromEmail=None, unreadCount=0,
        hasAttachments=False, hasDeadline=False,
        categoryId=None, categoryName=None, categoryColor=None,
    )


def map_conversation_detail(row: Any) -> ConversationDetail:
    """Map a Conversation (with threads.emails.memberships) to detail DTO."""
    base = map_conversation_summary(row)
    all_emails = sorted(
        (e for t in (row.threads or []) for e in (t.emails or [])),
        key=lambda e: e.receivedAt or 0,
    )
    total = len(all_emails)
    messages = [
        ConversationMessage(
            id=e.id, emailId=e.id, threadId=e.threadId,
            providerThreadId=e.providerThreadId, fromName=e.fromName,
            fromEmail=e.fromEmail, subject=e.subject, snippet=e.snippet,
            receivedAt=e.receivedAt.isoformat() if e.receivedAt else None,
            isRead=e.isRead, isImportant=e.isImportant,
            hasAttachment=e.hasAttachment,
            messageType=detect_message_type(e.subject, i, total),
            isLatest=i == total - 1,
            categoryId=(e.memberships[0].category.id if e.memberships else None),
            categoryName=(e.memberships[0].category.name if e.memberships else None),
            categoryColor=(e.memberships[0].category.color if e.memberships else None),
        )
        for i, e in enumerate(all_emails)
    ]
    return ConversationDetail(
        **base.model_dump(by_alias=True),
        messages=messages, changes=_detect_changes(all_emails),
        aiSummary=row.aiSummary,
    )
